### PR TITLE
Add correct response on incorrect JSON

### PR DESCRIPTION
**Pull request by Morgan-iv**
Implement correct response on requests with incorrect JSON instead of just panic. Errors in `body_to_request` are propagated to response if they occurred before request deserialization

### CODE REVIEW

**Comment by Morgan-iv** [comment]
[clabot:check]

File: src/transport/http/server/background.rs [hunk 1]
```diff
@@ -166,8 +166,14 @@ async fn process_request(
 
             let json_body = match body_to_request(request.into_body()).await {
                 Ok(b) => b,
-                Err(_) => {
-                    unimplemented!() // TODO:
+                Err(e) => {
+                    use std::io::ErrorKind::*;
+                    match (e.kind(), e.into_inner()) {
+                        (InvalidData, _) => return response::parse_error(),
+                        (UnexpectedEof, _) => return response::parse_error(),
```
Comment (tomaka) [inline src/transport/http/server/background.rs]: This is very subjective, but could you remove importing `*`, and imports within the body of functions?

```suggestion
                    match (e.kind(), e.into_inner()) {
                        (io::ErrorKind::InvalidData, _) => return response::parse_error(),
                        (io::ErrorKind::UnexpectedEof, _) => return response::parse_error(),
```

File: src/transport/http/server/response.rs [hunk 2]
```diff
@@ -63,6 +63,23 @@ pub fn unsupported_content_type() -> hyper::Response<hyper::Body> {
     )
 }
 
+/// Create a response for invalid JSON in request
+pub fn parse_error() -> hyper::Response<hyper::Body> {
+    use crate::common::*;
```
Comment (tomaka) [inline src/transport/http/server/response.rs]:
```suggestion
```

File: src/transport/http/server/response.rs [hunk 3]
```diff
@@ -63,6 +63,23 @@ pub fn unsupported_content_type() -> hyper::Response<hyper::Body> {
...
+    hyper::Response::builder()
+        .status(hyper::StatusCode::OK)
+        .header("Content-type", "application/json")
+        .body(hyper::Body::from(
+            serde_json::to_string(&Output::Failure(Failure {
+                jsonrpc: Version::V2,
+                error: Error::parse_error(),
+                id: Id::Null,
```
Comment (tomaka) [inline src/transport/http/server/response.rs]:
```suggestion
            serde_json::to_string(&common::Output::Failure(common::Failure {
                jsonrpc: common::Version::V2,
                error: common::Error::parse_error(),
                id: common::Id::Null,
```

**Comment by Morgan-iv** [comment]
I added your changes, but now Travis CI tests fail for some reason 😔

**Comment by Morgan-iv** [comment]
I returned to this PR a few days later, found what was the reason for the Travis build error and fixed it. Now it builds fully correctly

**Pull request merged**